pkg/kubelet/nodelease/controller_test.go

```diff
@@ -195,3 +200,85 @@ func TestNewLease(t *testing.T) {
 		})
 	}
 }
+
+func TestRetryUpdateLease(t *testing.T) {
+	node := &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+			UID:  types.UID("foo-uid"),
+		},
+	}
+	gr := schema.GroupResource{Group: "v1", Resource: "lease"}
+	noConnectionUpdateErr := apierrors.NewServerTimeout(gr, "put", 1)
+	optimistcLockUpdateErr := apierrors.NewConflict(gr, "lease", fmt.Errorf("conflict"))
+	cases := []struct {
+		desc                       string
+		updateReactor              func(action clienttesting.Action) (bool, runtime.Object, error)
+		getReactor                 func(action clienttesting.Action) (bool, runtime.Object, error)
+		onRepeatedHeartbeatFailure func()
+		expectErr                  error
+	}{
+		{
+			desc: "no errors",
+			updateReactor: func(action clienttesting.Action) (bool, runtime.Object, error) {
+				return true, &coordv1beta1.Lease{}, nil
+			},
+			getReactor:                 nil,
+			onRepeatedHeartbeatFailure: nil,
+			expectErr:                  nil,
+		},
+		{
+			desc: "connection errors",
+			updateReactor: func(action clienttesting.Action) (bool, runtime.Object, error) {
+				return true, nil, noConnectionUpdateErr
+			},
+			getReactor:                 nil,
+			onRepeatedHeartbeatFailure: nil,
+			expectErr:                  fmt.Errorf("failed %d attempts to update node lease", maxUpdateRetries),
+		},
+		{
+			desc: "optimistic lock errors",
+			updateReactor: func() func(action clienttesting.Action) (bool, runtime.Object, error) {
+				i := 0
+				return func(action clienttesting.Action) (bool, runtime.Object, error) {
+					i++
+					switch i {
+					case 1:
+						return true, nil, noConnectionUpdateErr
+					case 2:
+						return true, nil, optimistcLockUpdateErr
+					default:
+						return true, &coordv1beta1.Lease{}, nil
+					}
+				}
+			}(),
+			getReactor: func(action clienttesting.Action) (bool, runtime.Object, error) {
+				return true, &coordv1beta1.Lease{}, nil
+			},
+			onRepeatedHeartbeatFailure: func() { t.Fatalf("onRepeatedHeartbeatFailure called") },
+			expectErr:                  nil,
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.desc, func(t *testing.T) {
+			cl := fake.NewSimpleClientset(node)
+			if tc.updateReactor != nil {
+				cl.PrependReactor("update", "leases", tc.updateReactor)
+			}
+			if tc.getReactor != nil {
+				cl.PrependReactor("get", "leases", tc.getReactor)
+			}
+			c := &controller{
+				clock:                      clock.NewFakeClock(time.Now()),
+				client:                     cl,
+				leaseClient:                cl.CoordinationV1beta1().Leases(corev1.NamespaceNodeLease),
+				holderIdentity:             node.Name,
+				leaseDurationSeconds:       10,
+				onRepeatedHeartbeatFailure: tc.onRepeatedHeartbeatFailure,
+			}
+			if err := c.retryUpdateLease(nil); fmt.Sprintf("%v", err) != fmt.Sprintf("%v", tc.expectErr) {
```

Comment:
Why we need fmt.Sprintf?

Please fix so that we could simply compare errors.

Reply:
This is a hacky way to compare errors - normally you use Error(), however this will not work with nil. To avoid nil checks etc, using Spritnf will cause to call Error() if err != nil, or just returns <nil>.

Assuming that we don't really need to compare errors, but rather check if error was returned or not, I can change ecpectErr to bool and verify of err != nil. WDYT?

Reply:
> Assuming that we don't really need to compare errors, but rather check if error was returned or not, I can change ecpectErr to bool and verify of err != nil. WDYT?

+1

Reply:
Done.